aten/src/ATen/core/dispatch/DispatchKeyExtractor.h

```diff
@@ -61,7 +61,7 @@ namespace detail {
         ts = ts | x->key_set();
       }
     }
-    void operator()(at::ArrayRef<at::Tensor> xs) {
+    void operator()(at::ITensorList xs) {
```

Comment:
Hmm... does stuff break if we leave this as `ArrayRef<Tensor>`?

I might be wrong, but I think that this code lives entirely in unboxed world, which means that we know the underlying this is an `ArrayRef<Tensor>`. This is also pretty hot-path code in the dispatcher's runtime dispatch logic.

Reply:
I think we could leave the `ArrayRef<Tensor>` implementation alongside the `ITensorList` one (not sure that was what you meant). That's because with the introduction of `ITensorList`, boxed lists of tensors might get to this code path.

Reply:
I don't think that boxed lists of tensor would actually reach that code path - the `DispatchKeyExtractor` is responsible for merging all of the tensor-like arguments from an op into a single `DispatchKeySet`, and it has separate functions for doing that in the boxed world (`getDispatchKeySetBoxed()`) and the unboxed world (`getDispatchKeySetUnboxed`).

The boxed version is [here](https://github.com/pytorch/pytorch/blob/5e6f29661254e5ebc97fac3d829e5d455cda5864/aten/src/ATen/core/dispatch/DispatchKeyExtractor.h#L144), and it doesn't go through the `MultiDispatchKeySet` struct. So we can assume that only `ArrayRef<Tensor>` objects will ever get here (and I don't think we'll need an `ITensorList` template specialization.

Reply:
Hmm. That was true. But, if I understand it correctly, `ITensorList` changes that. Consider, for example, the case where we somehow call `at::stack(boxed)` from the boxed world:

1. `boxed` (`IValue`) is converted to `List<Tensor>` (`ivalue_to_arg`), and then to `ITensorList` (let's call it `tlist`)
2. `at::stack` gets dispatched to `at::native::stack` function
3. `at::native::stack` calls `at::cat(tlist)`

In (3), we enter the dispatcher in the unboxed code path. However, one of its argumes is `tlist`, which is basically a pointer to a boxed representation `List<Tensor>`.

Reply:
Hmmm. Good question. Maybe it's possible, but wouldn't that be the same as not having `ITensorList` at all, while cloning functions for `ArrayRef<Tensor>` and `List<Tensor>`? Anyway, I think the biggest challenge would be to duplicating every function that has a list of tensor as their parameters.